### PR TITLE
Keep `ChainWorkerActor` running in an LRU cache

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

The worker spawns a `ChainWorkerActor` to handle each request. This is inefficient because it loads from storage the chain state every time. This should be cached in memory in order to avoid paying the overhead of deserializing the chain state every time.

## Proposal

Keep an LRU cache with some running `ChainWorkerActor`s, each responsible for a different chain.

The size of the cache should be configurable by the validators, because there's a tradeoff between chain execution latency and memory usage.

## Test Plan

This is an internal refactor, so existing tests should catch any regressions.

I manually ran the Crowd-Funding `collect_pledges` integration test and collected some metrics before and after this PR.

Before:

```
# HELP linera_load_view The metric counting how often a view is read from storage
# TYPE linera_load_view counter
linera_load_view{base_key="001db1936dad0717597a7743a8353c9c0191c14c3a129b258e9743aec2b4f05d03",type="ChainStateView"} 21
linera_load_view{base_key="0036137b1ddac30eb9d0d4821a1560e66795dbdad6dbf4b708a8d81ca3df866e1b",type="ChainStateView"} 67
linera_load_view{base_key="00a393137daba303e8b561cb3a5bff50efba1fb7f24950db28f1844b7ac2c1cf27",type="ChainStateView"} 37
linera_load_view{base_key="00ae41b40b288a1e7ed064e2ff749a9ce3e780a5742dca074e6015e77e9dd373f8",type="ChainStateView"} 37
linera_load_view{base_key="00be20093606a7296fbda537060becfecc62b5441fa784b3d26d6742152a80a1f9",type="ChainStateView"} 61
linera_load_view{base_key="00e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65",type="ChainStateView"} 36
linera_load_view{base_key="00fc9384defb0bcd8f6e206ffda32599e24ba715f45ec88d4ac81ec47eb84fa111",type="ChainStateView"} 21
linera_load_view{base_key="00feeeed172c74027e2675311c4ec3239dbe8e4b4fcf46b12e6f775523a002d64e",type="ChainStateView"} 37
# HELP linera_load_chain_latency The latency to load a chain state
# TYPE linera_load_chain_latency histogram
linera_load_chain_latency_bucket{le="0.001"} 0
linera_load_chain_latency_bucket{le="0.0025"} 0
linera_load_chain_latency_bucket{le="0.005"} 0
linera_load_chain_latency_bucket{le="0.01"} 0
linera_load_chain_latency_bucket{le="0.025"} 2
linera_load_chain_latency_bucket{le="0.05"} 71
linera_load_chain_latency_bucket{le="0.1"} 313
linera_load_chain_latency_bucket{le="0.25"} 317
linera_load_chain_latency_bucket{le="0.5"} 317
linera_load_chain_latency_bucket{le="1"} 317
linera_load_chain_latency_bucket{le="+Inf"} 317
linera_load_chain_latency_sum 19.172113999999997
linera_load_chain_latency_count 317
```

After:
```
# HELP linera_load_view The metric counting how often a view is read from storage
# TYPE linera_load_view counter
linera_load_view{base_key="001db1936dad0717597a7743a8353c9c0191c14c3a129b258e9743aec2b4f05d03",type="ChainStateView"} 1
linera_load_view{base_key="0036137b1ddac30eb9d0d4821a1560e66795dbdad6dbf4b708a8d81ca3df866e1b",type="ChainStateView"} 1
linera_load_view{base_key="00a393137daba303e8b561cb3a5bff50efba1fb7f24950db28f1844b7ac2c1cf27",type="ChainStateView"} 1
linera_load_view{base_key="00ae41b40b288a1e7ed064e2ff749a9ce3e780a5742dca074e6015e77e9dd373f8",type="ChainStateView"} 1
linera_load_view{base_key="00be20093606a7296fbda537060becfecc62b5441fa784b3d26d6742152a80a1f9",type="ChainStateView"} 1
linera_load_view{base_key="00e476187f6ddfeb9d588c7b45d3df334d5501d6499b3f9ad5595cae86cce16a65",type="ChainStateView"} 2
linera_load_view{base_key="00fc9384defb0bcd8f6e206ffda32599e24ba715f45ec88d4ac81ec47eb84fa111",type="ChainStateView"} 1
linera_load_view{base_key="00feeeed172c74027e2675311c4ec3239dbe8e4b4fcf46b12e6f775523a002d64e",type="ChainStateView"} 1
# HELP linera_load_chain_latency The latency to load a chain state
# TYPE linera_load_chain_latency histogram
linera_load_chain_latency_bucket{le="0.001"} 0
linera_load_chain_latency_bucket{le="0.0025"} 0
linera_load_chain_latency_bucket{le="0.005"} 0
linera_load_chain_latency_bucket{le="0.01"} 0
linera_load_chain_latency_bucket{le="0.025"} 0
linera_load_chain_latency_bucket{le="0.05"} 5
linera_load_chain_latency_bucket{le="0.1"} 9
linera_load_chain_latency_bucket{le="0.25"} 9
linera_load_chain_latency_bucket{le="0.5"} 9
linera_load_chain_latency_bucket{le="1"} 9
linera_load_chain_latency_bucket{le="+Inf"} 9
linera_load_chain_latency_sum 0.492632
linera_load_chain_latency_count 9
```

In summary, the time spent loading chains during the test went from 19.2ms to 0.5ms. And that was possible because this PR reduced the number of chain load events from 317 to 9.

## Release Plan

This is an internal refactor, so nothing is needed.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-core/src/worker.rs [hunk 1]
```diff
@@ -270,7 +277,11 @@ impl From<linera_chain::ChainError> for WorkerError {
 
 /// State of a worker in a validator or a local node.
 #[derive(Clone)]
-pub struct WorkerState<StorageClient> {
+pub struct WorkerState<StorageClient>
```
Comment (ma2bd) [inline linera-core/src/worker.rs:280]: Can we add a parameter `(Chain)WorkerRequest` and avoid the `where` clauses?

File: linera-core/src/client.rs [hunk 2]
```diff
@@ -199,7 +207,11 @@ impl MessagePolicy {
 /// * The chain being operated is called the "local chain" or just the "chain".
 /// * As a rule, operations are considered successful (and communication may stop) when
 /// they succeeded in gathering a quorum of responses.
-pub struct ChainClient<ValidatorNodeProvider, Storage> {
+pub struct ChainClient<ValidatorNodeProvider, Storage>
+where
+    Storage: linera_storage::Storage,
+    ViewError: From<Storage::ContextError>,
+{
```
Comment (afck) [inline linera-core/src/client.rs:214]: Those might not be needed on the struct definitions, only on the `impl` blocks?

File: linera-core/src/worker.rs [hunk 3]
```diff
@@ -286,12 +297,22 @@ pub struct WorkerState<StorageClient> {
     delivery_notifiers: Arc<Mutex<DeliveryNotifiers>>,
     /// The set of spawned [`ChainWorkerActor`] tasks.
     chain_worker_tasks: Arc<Mutex<JoinSet<()>>>,
+    /// The cache of running [`ChainWorkerActor`]s.
+    chain_workers: Arc<Mutex<LruCache<ChainId, ChainActorEndpoint<StorageClient>>>>,
 }
 
+/// The sender endpoint for [`ChainWorkerRequest`]s.
+type ChainActorEndpoint<StorageClient> =
+    mpsc::UnboundedSender<ChainWorkerRequest<<StorageClient as Storage>::Context>>;
```
Comment (afck) [inline linera-core/src/worker.rs:306]: Consider creating an issue to revisit unbounded senders in general: backpressure might be better than the risk of running out of memory.

**Review by afck — APPROVED** [review]

File: linera-core/src/client.rs (hunk 2 above)
Comment (jvff) [reply linera-core/src/client.rs:215]: Unfortunately it is, because of the `where` clause in `ChainStateView` (see discussion in previous comment).

File: linera-core/src/worker.rs (hunk 3 above)
Comment (jvff) [reply linera-core/src/worker.rs:306]: Good idea, I opened #2176 for that :+1:

**Review by ma2bd — APPROVED** [review]

**Review by ma2bd — APPROVED** [review]
Epic PR! This is going to unlock so many optimizations

**Review by andresilva91 — COMMENTED** [review]
This is awesome!
Added via Giphy

File: linera-core/src/worker.rs [hunk 4]
```diff
@@ -733,10 +749,45 @@ where
...
+                    self.storage.clone(),
+                    self.recent_hashed_certificate_values.clone(),
+                    self.recent_hashed_blobs.clone(),
+                    chain_id,
+                    &mut *self.chain_worker_tasks.lock().await,
+                )
+                .await?,
+            );
+        }
+
```
Comment (andresilva91) [inline linera-core/src/worker.rs:775]: Should we also promote `chain_id` here on the `LruCache`? 🤔
